### ISSUE TITLE
SparseSemiStructuredTensors are constructed differently from the original dense ones

### ISSUE THREAD

**Issue by irishev** [pytorch/pytorch#109175]
### 🐛 Describe the bug

I'm working on semi-structured sparsity using the recently introduced features of SparseSemiStructuredTensors.

However, I have encountered inconsistent results between the dense (i.e., HalfTensor) and sparse (i.e., SparseSemiStructuredTensors) formats of 2:4 pruned weights.
My models almost always perform worse when using sparse formats than dense formats.
Moreover, when I reconstructed a dense tensor from a sparse tensor, I found several weights were indexed differently.
There may exist some errors in encoding indices resulting in these inconsistent behaviors.

Here's a minimal code for reproducing the problem.

```
import torch
from torch.sparse import to_sparse_semi_structured

a = torch.randn(3072, 768).cuda().half()/100
pruning_inds = a.abs().view(-1, 4).argsort(dim=1)[:, :2]

a.view(-1, 4).scatter_(1, pruning_inds, torch.zeros_like(pruning_inds).half())

b = to_sparse_semi_structured(a).to_dense()

print(((a-b)**2).sum())
print(a[a!=b][:32])
print(b[a!=b][:32])
```

Here's my result of running the above code. It can be clearly found that the remaining weights are positioned wrong.
```
tensor(239.8750, device='cuda:0', dtype=torch.float16)
tensor([ 0.0000, -0.0087,  0.0000,  0.0162,  0.0000, -0.0073,  0.0163,  0.0000,                                                                                                                                                       
         0.0000,  0.0155,  0.0000, -0.0177,  0.0128,  0.0000,  0.0000, -0.0084,                                                                                                                                                       
        -0.0058,  0.0000,  0.0000,  0.0124,  0.0000,  0.0093,  0.0000,  0.0060,                                                                                                                                                       
         0.0078,  0.0145,  0.0000, -0.0056,  0.0000,  0.0157,  0.0146,  0.0000],                                                                                                                                                      
       device='cuda:0', dtype=torch.float16)
tensor([-0.0087,  0.0000,  0.0162,  0.0000, -0.0073,  0.0000,  0.0000,  0.0163,                                                                                                                                                       
         0.0155,  0.0000, -0.0177,  0.0000,  0.0000,  0.0128, -0.0084,  0.0000,                                                                                                                                                       
         0.0000, -0.0058,  0.0124,  0.0000,  0.0093,  0.0000,  0.0060,  0.0000,
         0.0000,  0.0078,  0.0145,  0.0000, -0.0056,  0.0000,  0.0157,  0.0146],
       device='cuda:0', dtype=torch.float16)
```

Thanks.

### Versions

```
Collecting environment information...                                                                                                                                                                                                 
PyTorch version: 2.2.0a0+gitb193f29                                                                                                                                                                                                   
Is debug build: False                                                                                                                                                                                                                 
CUDA used to build PyTorch: 12.2                                                                                                                                                                                                      
ROCM used to build PyTorch: N/A                                                                                                                                                                                                       

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.31

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-162-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA RTX A5000
GPU 2: NVIDIA RTX A5000
GPU 3: NVIDIA RTX A5000

Nvidia driver version: 535.86.10
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.4                                                                                                                                                                                           
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.4                                                                                                                                                                                 
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.4                                                                                                                                                                                 
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.4                                                                                                                                                                                 
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             36
On-line CPU(s) list:                0-35
Thread(s) per core:                 2
Core(s) per socket:                 18
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Core(TM) i9-10980XE CPU @ 3.00GHz                                                                                                                                                        
Stepping:                           7                                                                                                                                                                                                 
CPU MHz:                            1200.079                                                                                                                                                                                          
CPU max MHz:                        4800.0000                                                                                                                                                                                         
CPU min MHz:                        1200.0000                                                                                                                                                                                         
BogoMIPS:                           6000.00                                                                                                                                                                                           
Virtualization:                     VT-x                                                                                                                                                                                              
L1d cache:                          576 KiB                                                                                                                                                                                           
L1i cache:                          576 KiB                                                                                                                                                                                           
L2 cache:                           18 MiB
L3 cache:                           24.8 MiB
NUMA node0 CPU(s):                  0-35
Vulnerability Gather data sampling: Mitigation; Microcode 
Vulnerability Itlb multihit:        KVM: Mitigation: Split huge pages
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512_vnni md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] pytorch-triton==2.1.0+e6216047b8
[pip3] torch==2.2.0a0+gitb193f29
[pip3] torchaudio==2.2.0.dev20230906+cu121
[pip3] torchvision==0.16.0.dev20230906+cu121
[conda] magma-cuda121             2.6.1                         1    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46343  
[conda] mkl-include               2023.1.0         h06a4308_46343  
[conda] numpy                     1.25.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+e6216047b8          pypi_0    pypi
[conda] torch                     2.2.0a0+gitb193f29           dev_0    <develop>
[conda] torchaudio                2.2.0.dev20230906+cu121          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230906+cu121          pypi_0    pypi
```

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

**Comment by jcaip** [comment]
@alexsamardzic forcing CUTLASS solves this issue from my end, so that's good. But that means cuSPARSELt and CUTLASS compression functions are not identical? That seems weird to me. I wrote a short script to test this and it looks like this is indeed the case:

```
import torch
from torch.sparse import to_sparse_semi_structured, SparseSemiStructuredTensor

def test(force_cutlass):
    print(f"TESTING CUTLASS: {force_cutlass}")
    print("-------------------")
    SparseSemiStructuredTensor._FORCE_CUTLASS = force_cutlass
    torch.manual_seed(52)
    a = torch.randn(256, 256).cuda().half()/100

    pruning_inds = a.abs().view(-1, 4).argsort(dim=1)[:, :2]

    a.view(-1, 4).scatter_(1, pruning_inds, torch.zeros_like(pruning_inds).half())
    a_sparse = to_sparse_semi_structured(a)
    print(a_sparse.indices())
    b = a_sparse.to_dense()

    print(((a-b)**2).sum())
    print(a[a!=b][:32])
    print(b[a!=b][:32])

    return a_sparse.indices().detach()

cutlass_indicies = test(True)
cusparselt_indicies = test(False)

torch.set_printoptions(profile="full")
print("different indices at locations:")
print(torch.nonzero((cutlass_indicies - cusparselt_indicies)**2))
```

I see the following output, looking at the indices of the differences, they definitely seem to follow a pattern. 
```
TESTING CUTLASS: True
-------------------
/home/jessecai/local/A/pytorch/torch/sparse/semi_structured.py:86: UserWarning: The PyTorch API of SparseSemiStructuredTensor is in prototype stage and will change in the near future. Please open a Github issue for features requests and see our documentation on the torch.sparse module for further information about the project.
  warnings.warn(
tensor(0., device='cuda:0', dtype=torch.float16)
tensor([], device='cuda:0', dtype=torch.float16)
tensor([], device='cuda:0', dtype=torch.float16)

TESTING CUTLASS: False
-------------------
tensor(3.4609, device='cuda:0', dtype=torch.float16)
tensor([-0.0067,  0.0000,  0.0000, -0.0024,  0.0000,  0.0205,  0.0137, -0.0202,
         0.0064,  0.0000,  0.0000, -0.0069, -0.0149, -0.0055,  0.0000, -0.0085,
         0.0000,  0.0000,  0.0175,  0.0169, -0.0217,  0.0000,  0.0054, -0.0018,
         0.0000,  0.0000,  0.0032,  0.0000, -0.0020,  0.0052,  0.0220,  0.0000],
       device='cuda:0', dtype=torch.float16)
tensor([ 0.0000, -0.0067, -0.0024,  0.0000,  0.0205,  0.0137,  0.0000,  0.0000,
        -0.0202,  0.0064, -0.0069, -0.0149,  0.0000,  0.0000, -0.0055,  0.0000,
        -0.0085,  0.0175,  0.0000,  0.0000,  0.0169, -0.0217,  0.0000,  0.0054,
        -0.0018,  0.0032,  0.0000, -0.0020,  0.0000,  0.0000,  0.0052,  0.0220],
       device='cuda:0', dtype=torch.float16)

different indices at locations:
tensor([[ 16,   0],
        [ 16,   1],
        [ 16,   3],
        ...,
        [239,  12],
        [239,  14],
        [239,  15]], device='cuda:0')
1982 total differences 
```

**Comment by jcaip** [comment]
@irishev, thanks for flagging this. 

I think the cause of this discrepancy is using cuSPARSELt for compression but our `to_dense` method uses the CUTLASS format / method to convert back. However this should not be effective numerics in your model, as usually you should not be calling `to_dense` and instead calling torch.mm. 

When doing sparse matmul, the numerics should match exactly (we have tests for this). If you don't need `to_dense` explicitly, you can compare the output of `torch.mm(A_sparse, B)` and `torch.mm(A, B)`. This **should** be numerically equivalent, so please let me know if it looks off.

**Comment by irishev** [comment]
@alexsamardzic @jcaip Thank you for all your comments. Forcing the usage of CUTLASS solves all the issues.

But it still exists that the performance of pruned models becomes worse when using cuSPARSELt.
Specifically, I measured the fine-tuned accuracy of pruned BERT-base models by using the Hugging Face transformers.
The SST-2 accuracies are as follows:

```
# Without using sparse tensors
accuracy 0.9334862385321101
f1 0.9334806397306397
loss 0.45408626965114046

# With CUTLASS
accuracy 0.9334862385321101
f1 0.9334806397306397
loss 0.42543656485421316

# With cuSPARSELt
accuracy 0.6364678899082569
f1 0.6326410733816812
loss 1.15069580078125
```

After some investigations, I have found that the memory format of input tensors can cause problems.
I modified the forward of torch.nn.Linear and measured the SST-2 accuracies.

```
# F.linear(input, self.weight, self.bias) (Without modification)
accuracy 0.6364678899082569
f1 0.6326410733816812
loss 1.15069580078125

# F.linear(input, self.weight, self.bias).contiguous() 
accuracy 0.6353211009174312
f1 0.6313945265351915
loss 1.1506609235491072

# F.linear(input.contiguous(), self.weight, self.bias)
accuracy 0.9334862385321101
f1 0.9334806397306397
loss 0.42543275015694754

# F.linear(input.contiguous(), self.weight, self.bias).contiguous() 
accuracy 0.9323394495412844
f1 0.9323322412248801
loss 0.4254371098109654
```

I think that the reason might be that the transpose or permute functions during computing self-attention cannot be traced by cuSPARSELt for some reasons. As I found, the current version provides some combinations of memory formats for sparse-dense matmuls (e.g., A_sparse@B.T, A@B_sparse.T).

**Comment by jcaip** [comment]
Okay so I think it sounds like there are two issues here:

1. CUTLASS conversion/reverse being different from cuSPARSELt, for the reasons @alexsamardzic outlined above. I think renaming the conversions to be cutlass specific sounds good. We should also probably throw an error if `to_dense` is called and we used cuSPARSELt to compress the tensor. Let's ask NVIDIA too on slack if they know what the exact difference is. 

For installing cuSPARSELt you need to download the library from NVIDIA [here](https://developer.nvidia.com/cusparselt-downloads) and then you can build pytorch from source with some [additional flags set](https://github.com/pytorch/pytorch/issues/62153#issuecomment-1702888707). You should be able to see if it's installed by running `torch._cslt_compress` and seeing if that throws an exception. We do indeed default to CUTLASS when cuSPASRELt is not available. 

2. Separately cuSPARSELt expects the inputs to be in a certain memory format, and will return bad numerics if this is not the case, from the tests @irishev has run. I think I know the reason for this problem, which is that cuSPASRELt returns non-contiguous tensors. (I think CUTLASS does this as well, but not sure). I am pretty sure this problem arises in how we create the descriptors for non-contiguous activation tensors, which happens when we have multiple linear layers in a row. The non-contiguous output gets fed to the subsequent cuSPARSELt error and that's where we have our error. This is why we don't see numeric differences in our test, since they are all single-layer Linear models. 

I can debug this some more on my end. @irishev can you also try running with
```
from torch.sparse import to_sparse_semi_structured, SparseSemiStructuredTensor
SparseSemiStructuredTensor._FUSE_TRANSPOSE = True
```
This will fuse a contiguous call into the cuSPARSELt matmul, at the expense of some performance. But I think this should also fix your problem without explicit `.contiguous()` calls.

**Comment by irishev** [comment]
@jcaip:

I have tested your suggestion, but it does not fix the problem. It still requires ```.contiguous()``` calls to get accurate numerics.

**Comment by jcaip** [comment]
@irishev I am looking into this some more this week, can you share your code, specifically for the pruned BERT models?

**Comment by jcaip** [comment]
Closing as we've found out that cuSPARSELt has different conversion routines from CUTLASS, and have updated the code so that converting `to_dense` is disabled for cuSPASRELt sparse tensors.
